cellarium/ml/models/mu_linear.py

```diff
@@ -0,0 +1,300 @@
+    | :math:`n` | out_features   | in_features    | in_features    |
+    +-----------+----------------+----------------+----------------+
+
+    The maximal update parametrization for Adam and AdamW is defined by
+
+    +-----------+----------------+----------------+----------------+
+    |           | Input & Biases | Hidden         | Output         |
+    +-----------+----------------+----------------+----------------+
+    | :math:`a` | :math:`0`      | :math:`1`      | :math:`1`      |
+    +-----------+----------------+----------------+----------------+
+    | :math:`b` | :math:`0`      | :math:`-0.5`   | :math:`0`      |
+    +-----------+----------------+----------------+----------------+
+    | :math:`c` | :math:`0`      | :math:`0`      | :math:`0`      |
+    +-----------+----------------+----------------+----------------+
+    | :math:`d` | :math:`1`      | :math:`2`      | :math:`1`      |
+    +-----------+----------------+----------------+----------------+
+    | :math:`n` | out_features   | in_features    | in_features    |
+    +-----------+----------------+----------------+----------------+
+
+    Since in this implementation :math:`c` always equals 0, regular PyTorch optimizers
+    can be used.
+
+    **References:**
+
+    1. `Feature Learning in Infinite-Width Neural Networks (Yang et al.)
+       <https://arxiv.org/pdf/2011.14522.pdf>`_.
+    2. `Tensor Programs IVb: Adaptive Optimization in the ∞-Width Limit (Yang et al.)
+       <https://arxiv.org/pdf/2308.01814.pdf>`_.
+
+    Args:
+        in_features:
+            Size of each input sample.
+        out_features:
+            Size of each output sample.
+        bias:
+            If set to ``False``, the layer will not learn an additive bias.
+        layer:
+            Layer type.
+        optimizer:
+            Optimizer type.
+        weight_init_std:
+            The standard deviation of the weight initialization at base width.
+        bias_init_std:
+            The standard deviation of the bias initialization at base width.
+        lr_scale:
+            The learning rate scaling factor for the weight and the bias.
+        base_width:
+            The base width of the layer.
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool,
+        layer: Literal["input", "hidden", "output"],
+        optimizer: Literal["sgd", "adam", "adamw"],
+        weight_init_std: float = 1.0,
+        bias_init_std: float = 0.0,
+        lr_scale: float = 1.0,
+        base_width: int = 1,
+    ) -> None:
+        super().__init__()
+
+        # inf dim
+        bias_width = out_features  # TODO: what is the width of the bias of the output layer?
+        if layer == "input":
+            weight_width = out_features
+        else:
+            weight_width = in_features
+
+        # c = 0 for all layers by design
+
+        # a, b, d
+        if optimizer == "sgd":
```

Comment:
FYI, PEP-0636 adds pattern matching support (finally!):
https://peps.python.org/pep-0636/

You may want to use that in the future :)